src/state.rs

```diff
@@ -341,9 +358,42 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
+                    .get_cached_resolve_dns_for_hostname(&hostname)
+                    .is_none()
+                {
+                    let cached_resolve_dns = Arc::new(CachedAtomicBool {
+                        should_resolved_dns: AtomicBool::new(true),
+                        complete_resolved_dns: AtomicBool::new(false),
+                    });
+                    let cached_resolve_dns_clone = cached_resolve_dns.clone();
+                    state.set_cached_resolve_dns_for_hostname(
+                        hostname.to_owned(),
+                        cached_resolve_dns_clone,
+                    );
+                    if cached_resolve_dns
+                        .should_resolved_dns
+                        .load(Ordering::Relaxed)
+                    {
+                        cached_resolve_dns
+                            .should_resolved_dns
+                            .store(false, Ordering::Relaxed);
+                        Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                        cached_resolve_dns
+                            .complete_resolved_dns
+                            .store(true, Ordering::Relaxed);
+                    } else {
+                        while !cached_resolve_dns
+                            .complete_resolved_dns
+                            .load(Ordering::Relaxed)
+                        {
+                            tokio::time::sleep(Duration::from_millis(100)).await;
+                        }
```

Comment:
not an expert on dns, can you confirm what does core-dns do with such case

Reply:
Hi @hzxuzhonghu,  there should be many threads calling `load_balance_for_hostname` at the same time, and `resolve_on_demand_dns`, which will do DNS resolve task including using `lookup_ip` to get ip address of hostname, would be called many times if high concurrency requests are sent for specified hostname which need to be resolved.
The optimization is to make sure that the first request for hostname to resolve the DNS once and other same hostname request will wait here.
I will add some comments in the code. Thanks!

Reply:
Yeah but sleeping 100ms here is not appropriate

Reply:
This is a fairly common pattern https://pkg.go.dev/golang.org/x/sync/singleflight

Reply:
Greate usage for golang to use channel mechanism to implement singleflight pattern and `singleflight` has different context in different language. 😀
I will use the rust way to eliminate the sleep logic, thanks @howardjohn!

Reply:
Hi @howardjohn & @hzxuzhonghu, after some tries, I find out that `AtomicBool` is the easier way to implement the `singleflight` than MPSC channel in this case. Because I have to consider the async function in `load_balance_for_hostname` and async function calling `resolve_on_demand_dns`.
And I eliminate the sleep logic, then add the timeout for in other concurrency requests. Is this okay for now? Thanks!